pyogrio/raw.py

```diff
@@ -559,3 +559,173 @@ def write(
+        file, may be ignored by the driver, or may result in invalid files.
+    encoding : str, optional (default: None)
+        Only used for the .dbf file of ESRI Shapefiles. If not specified,
+        uses the default locale.
+    append : bool, optional (default: False)
+        If True, the data source specified by path already exists, and the
+        driver supports appending to an existing data source, will cause the
+        data to be appended to the existing records in the data source.
+        NOTE: append support is limited to specific drivers and GDAL versions.
+    dataset_metadata : dict, optional (default: None)
+        Metadata to be stored at the dataset level in the output file; limited
+        to drivers that support writing metadata, such as GPKG, and silently
+        ignored otherwise. Keys and values must be strings.
+    layer_metadata : dict, optional (default: None)
+        Metadata to be stored at the layer level in the output file; limited to
+        drivers that support writing metadata, such as GPKG, and silently
+        ignored otherwise. Keys and values must be strings.
+    metadata : dict, optional (default: None)
+        alias of layer_metadata
+    dataset_options : dict, optional
+        Dataset creation option (format specific) passed to OGR. Specify as
+        a key-value dictionary.
+    layer_options : dict, optional
+        Layer creation option (format specific) passed to OGR. Specify as
+        a key-value dictionary.
+    **kwargs
+        Additional driver-specific dataset or layer creation options passed
+        to OGR. pyogrio will attempt to automatically pass those keywords
+        either as dataset or as layer creation option based on the known
+        options for the specific driver. Alternatively, you can use the
+        explicit `dataset_options` or `layer_options` keywords to manually
+        do this (for example if an option exists as both dataset and layer
+        option).
+    """
+    if not HAS_ARROW_WRITE_API:
+        raise RuntimeError("GDAL>=3.8 required to write using arrow")
+
+    path = vsi_path(str(path))
+
+    if driver is None:
+        driver = detect_write_driver(path)
+
+    # verify that driver supports writing
+    if not ogr_driver_supports_write(driver):
+        raise DataSourceError(
+            f"{driver} does not support write functionality in GDAL "
+            f"{get_gdal_version_string()}"
+        )
+
+    # prevent segfault from: https://github.com/OSGeo/gdal/issues/5739
+    if append and driver == "FlatGeobuf" and get_gdal_version() <= (3, 5, 0):
+        raise RuntimeError(
+            "append to FlatGeobuf is not supported for GDAL <= 3.5.0 due to segfault"
+        )
+
+    if "promote_to_multi" in kwargs:
+        raise ValueError(
+            "The 'promote_to_multi' option is not supported when writing using Arrow"
+        )
+
+    if geometry_type is None:
+        raise ValueError("Need to specify 'geometry_type'")
```

Comment:
This is a bit of a dilemma. 
In theory, we can pass through the None, which gets translated to `wkbNone` on the GDAL side, which is documented as "non-standard, for pure attribute records". And that works fine for writing data without any geometry column.

However, if you forget to specify this and you _have_ a geometry the column (which is by far the typical use case), you get a generic *"RuntimeError: Error while writing batch to OGR layer."* error (i.e. the error we raise if `OGR_L_WriteArrowBatch` returns a non-success code). And that is not very helpful or clear that this is due not specifying the geometry type.

So for now I opted to require passing this (and so essentially not enabling to write no geometry column. You can still specify a geometry type, but then eg with GPKG you get an all-null geometry column)

Reply:
Maybe the error message in the described case would be more clear and useful if the inner error details were preserved? Something like "Cannot find OGR geometry field for Arrow array {geometry_name}".

Also, is `wkbNone` actually the most useful default here in the low-level interface? It means exactly "there are zero geometry fields", which should be uncommon. Maybe `wkbUnknown` (meaning "there is a single geometry field with no further restrictions") would be more useful?

Reply:
Would a 3-state variable make sense here, to enable writing without geometry?
- `""` (or some other arbitrary value to indicate unset): default, indicates not set by user and will raise exception
- `None`: provided by user to indicate there is no geometry to write
- `Point` ... `GeometryCollection`: valid geometry type